File: cpp/56. Merge Intervals.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vvi ans;       
        ans.push_back(intervals[0]);
        for(int i = 1; i < intervals.size(); i++) {
            vi a = ans.back();
            vi b = intervals[i];
            if(b[0] > a[1]) ans.push_back(b);
            else ans.back() = vi({a[0], max(a[1], b[1])});
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vvi intervals, ans;

    solution = Solution();
    intervals = {{1,3},{2,6},{8,10},{15,18}};
    ans = solution.merge(intervals);
    assert(ans==vvi({{1,6},{8,10},{15,18}}));

    solution = Solution();
    intervals = {{1,4},{4,5}};
    ans = solution.merge(intervals);
    assert(ans==vvi({{1,5}}));
}

File: python/56-Merge-Intervals.py
from typing import List

'''
https://blog.techbridge.cc/2020/01/16/leetcode-刷題-pattern-merge-intervals/

First we sort the intervals by start, then the conditions of each 2 intervals reduced to following 3 patterns.

1.
a: -----
b:    -----
a.end >= b.start so need to merge
==> merged: (a.start, max(a.end, b.end))

2. 
a: -----
b:  ---
a.end >= b.start so need to merge
==> merged: (a.start, max(a.end, b.end))

3. 
a: -----
b:        ---
a.end < b.start so do not need to merge
'''

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        "O(nlogn)"
        if len(intervals) == 0: return []

        intervals.sort(key=lambda x: x[0])
        result = [intervals[0]]

        for i in range(1, len(intervals)):
            a = result[-1]
            b = intervals[i]
            if a[1] >= b[0]:
                merged = [a[0], max(a[1], b[1])]
                result[-1] = merged
            else:
                result.append(b)

        return result


intervals = [[1,3],[2,6],[8,10],[15,18]]
result = Solution().merge(intervals)
assert result == [[1,6],[8,10],[15,18]]

intervals = [[1,4],[4,5]]
result = Solution().merge(intervals)
assert result == [[1,5]]